Optimizer: add some floating point optimizations

eeckstein opened the pull request:
The goal is to completely constant fold literals of custom floating point types which use the generic initializers of the stdlib's `BinaryFloatingPoint` protocol.

Two changes were necessary to make this happen:
* add constant folding for some floating point intrinsics
* force inlining of some floating point functions in the stdlib, if their arguments are constant

It turned out that constant folding the `bitcast` intrinsic broke diagnosing floating point literals which end up as "infinity".
So, I had to fix this as well. For details see the commit messages.

rdar://147278006

eeckstein committed:
add a semantic attribute "optimize.sil.inline.constant.arguments"

This forces inlining the annotated function if its arguments are constant.

eeckstein committed:
stdlib: annotate some floating point functions with "optimize.sil.inline.constant.arguments"

eeckstein committed:
ConstantFolding: add constant folding for some floating point intrinsics

* `sitofp` signed integer to floating point
* `rint` round floating point to integral
* `bitcast` between integer and floating point

Constant folding `bitcast`s also made it necessary to rewrite constant folding for Nan and inf values, because the old code explicitly checked for `bitcast` intrinsics.
Relying on constant folded `bitcast`s makes the new version much simpler.

It is important to constant fold these intrinsics already in SIL because it enables other optimizations.

eeckstein committed:
tests: add an end-to-end to check complete constant folding of generic floating point literal construction.

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci apple silicon benchmark

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

rjmccall commented on test/SILOptimizer/diagnostic_constant_propagation_floats.swift:
```diff
@@ -77,13 +77,13 @@ func testFPToIntConversion() {
   let f1: Float = 1E38
   _blackHole(f1)
-  let f2: Float = 1E39 // expected-warning {{'1E39' overflows to inf during conversion to 'Float'}}
+  let f2: Float = 1E39 // expected-warning {{'1E39' overflows to inf because its magnitude exceeds the limits of a float literal}}
```
I agree with Steve: we should not talk about this as a limitation of floating point literals. There is no alternative way of computing this value that would yield a `Float` of magnitude `1E39`, because that is outside the range of `Float`. The diagnostic needs to say that. If you want to talk about floating-point literals in the diagnostic, consider something like "floating-point literal '1E39' overflows to inf because its magnitude exceeds the limits of `Float`".

eeckstein replied:
Oh, I think I misunderstood - I was referring to the "during conversion" part of the message.
Yes, it makes sense to include the actual floating point type name.

eeckstein replied:
I found an easier way to fix this diagnostic emission, with no changes to the wording.

eeckstein committed:
ConstantFolding: only print a warning for infinity floating point literals if it has an AST node

This avoids false alarms for floating point literals which are imported from the `var infinity` getter of floating point types.

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test windows

eeckstein commented:
@swift-ci smoke test macos
